File: tests/functional/Bots/DotNet/Consumers/CodeFirst/SimpleHostBot/SkillsConfiguration.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System;
using System.Collections.Generic;
using Microsoft.Bot.Builder.Skills;
using Microsoft.Extensions.Configuration;

namespace Microsoft.BotFrameworkFunctionalTests.SimpleHostBot
{
    /// <summary>
    /// A helper class that loads Skills information from configuration.
    /// </summary>
    public class SkillsConfiguration
    {
        /// <summary>
        /// Initializes a new instance of the <see cref="SkillsConfiguration"/> class to load skills information from configuration.
        /// </summary>
        /// <param name="configuration">The configuration properties.</param>
        public SkillsConfiguration(IConfiguration configuration)
        {
            var section = configuration?.GetSection("BotFrameworkSkills");
            var skills = section?.Get<BotFrameworkSkill[]>();
            if (skills != null)
            {
                foreach (var skill in skills)
                {
                    Skills.Add(skill.Id, skill);
                }
            }

            var skillHostEndpoint = configuration?.GetValue<string>(nameof(SkillHostEndpoint));
            if (!string.IsNullOrWhiteSpace(skillHostEndpoint))
            {
                SkillHostEndpoint = new Uri(skillHostEndpoint);
            }
        }

        /// <summary>
        /// Gets the URI representing the endpoint of the host bot.
        /// </summary>
        /// <value>
        /// The URI representing the endpoint of the host bot.
        /// </value>
        public Uri SkillHostEndpoint { get; }

        /// <summary>
        /// Gets the key-value pairs with the skills bots.
        /// </summary>
        /// <value>
        /// The key-value pairs with the skills bots.
        /// </value>
        public Dictionary<string, BotFrameworkSkill> Skills { get; } = new Dictionary<string, BotFrameworkSkill>(StringComparer.OrdinalIgnoreCase);
    }
}

File: tests/functional/Bots/JavaScript/Consumers/CodeFirst/SimpleHostBot/skillsConfiguration.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

/**
 * A helper class that loads Skills information from configuration.
 */
class SkillsConfiguration {
  constructor () {
    this.skillsData = {};

    const skillVariables = Object.keys(process.env).filter(prop => prop.startsWith('skill_'));

    for (const val of skillVariables) {
      const names = val.split('_');
      const id = names[1];
      const attr = names[2];
      let propName;
      if (!(id in this.skillsData)) {
        this.skillsData[id] = { id: id };
      }
      switch (attr.toLowerCase()) {
        case 'appid':
          propName = 'appId';
          break;
        case 'endpoint':
          propName = 'skillEndpoint';
          break;
        case 'group':
          propName = 'group';
          break;
        default:
          throw new Error(`[SkillsConfiguration]: Invalid environment variable declaration ${val}`);
      }

      this.skillsData[id][propName] = process.env[val];
    }

    this.skillHostEndpointValue = process.env.SkillHostEndpoint;
    if (!this.skillHostEndpointValue) {
      throw new Error('[SkillsConfiguration]: Missing configuration parameter. SkillHostEndpoint is required');
    }
  }

  get skills () {
    return this.skillsData;
  }

  get skillHostEndpoint () {
    return this.skillHostEndpointValue;
  }
}

module.exports.SkillsConfiguration = SkillsConfiguration;
